pennylane/ops/qubit/matrix_ops.py

```diff
@@ -27,6 +28,44 @@
 from pennylane.wires import Wires
 
 
+_walsh_hadamard_matrix = np.array([[1, 1], [1, -1]]) / 2
+
+
+def _walsh_hadamard_transform(D, n=None):
```

Comment:
I'd never heard of this before. This is pretty cool! (although I don't understand it 😅 ). Perhaps some references to documentation for developers could be useful

Reply:
So you heard of the `Hadamard` gate, which is basically the same transform. I'm not entirely sure about when the name "Walsh" should be added, but there is the so-called ["fast Walsh-Hadamard transform"](https://en.wikipedia.org/wiki/Fast_Walsh%E2%80%93Hadamard_transform) which plays a few tricks in order to _not_ do the matrix-vector multiplication $H^{\otimes n} \cdot v$, where $H$ is the single-qubit Hadamard matrix.

However, there is a plot twist: Performing the `tensordot` as is done in this function actually scales exactly the same way as the fast Walsh-Hadamard transform, just because we don't use large matrices but rather have the small `_wash_hadamard_matrix` at each step and use a `(2, 2, ... 2)` -shaped tensor for the vector $v$ above.
@trbromley noticed this only very recently, I should update the docstring :)